FM Box Display Automation

labkey-danield opened the pull request:
#### Rationale
In the FreezerManager app there is a breadcrumb component that is used for navigation. It is a combination of links, texts and dropdown menus. It is possible that this control will be used outside of the Inventory module so it should go into the higher level testAutomation library.

#### Related Pull Requests
* https://github.com/LabKey/sampleManagement/pull/420

#### Changes
* Implement the FMBreadCrumbs component.

**Bread Crumb Component**
At it's most basic a bread crumb control could look like this:
![image](https://user-images.githubusercontent.com/12738200/100028147-1c246b00-2da3-11eb-9118-ee9c21b0b67e.png)

This is just a `<ol>` element with several `<li><a>` elements.

In a more complex layout a breadcrumb can look like this:
![image](https://user-images.githubusercontent.com/12738200/100028371-a076ee00-2da3-11eb-84d6-1a968d867977.png)

This is still an `<ol>` with `<li>` elements, but now the `<li>` may contain a span (plain text) or a link. And some of the breadcrumbs will have a dropdown menu.

This test component will expose interaction with elements in the `<ol>` and give any the dropdown menu an element may have.

labkey-danield committed:
Merge branch 'develop' into fb_fmBoxDisplayAutomation

labkey-danield committed:
Merge branch 'develop' into fb_fmBoxDisplayAutomation

labkey-danield committed:
Create the test component that will be used to interact with the breadcrumb navigation element on the various freezer manager pages.

labkey-tchad commented on src/org/labkey/test/components/glassLibrary/navigation/FMBreadCrumbs.java:
```diff
@@ -0,0 +1,210 @@
+import java.util.List;
+
+/**
+ * <p>
+ *     This is a navigational component for the breadcrumb.ts element <i>I think</i>.
+ * </p>
```
I think this is specific to Freezer Manager ([LocationBreadcrumb.tsx](https://github.com/LabKey/inventory/blob/develop/src/client/FreezerManager/pages/freezers/navigation/LocationBreadcrumb.tsx)). If so the test component should go in the inventory module with it.

labkey-danield replied:
Moved.

labkey-tchad commented on src/org/labkey/test/components/glassLibrary/navigation/FMBreadCrumbs.java:
```diff
@@ -0,0 +1,210 @@
+    final WebElement _breadCrumbsElement;
+    final WebDriver _driver;
+
+    public FMBreadCrumbs(WebDriver driver)
+    {
+        this(Locators.component.findElement(driver), driver);
+    }
```
This sort of duplicates a globally scoped `FMBreadCrumbsFinder`. Consider removing this constructor in favor of using the Finder.

labkey-danield replied:
Yep. A copy and paste mistake.
Removed.

labkey-tchad commented on src/org/labkey/test/components/glassLibrary/navigation/FMBreadCrumbs.java:
```diff
@@ -0,0 +1,210 @@
+     */
+    public List<String> getBreadCrumbs()
+    {
+        List<String> breadCrumbs = new ArrayList<>();
+        for(WebElement crumb : getBreadCrumbElements())
+        {
+            breadCrumbs.add(crumb.getText());
+        }
+
+        return breadCrumbs;
```
We have a helper for this sort of thing:
```suggestion
        return getWrapper().getTexts(getBreadCrumbElements());
```

labkey-danield replied:
Changed.

labkey-tchad commented on src/org/labkey/test/components/glassLibrary/navigation/FMBreadCrumbs.java:
```diff
@@ -0,0 +1,210 @@
+    {
+        WebElement breadCrumb = elementCache().breadCrumb(breadCrumbText);
+        DropdownButtonGroup dropDownMenu = elementCache().dropDownMenu(breadCrumb);
+        if(!dropDownMenu.isExpanded())
+        {
+            dropDownMenu.expand();
+        }
+        return elementCache().dropDownMenu(breadCrumb).getMenuItemTexts();
```
It looks like `getMenuItemTexts` will open the menu if needed.
```suggestion
        return elementCache().dropDownMenu(breadCrumb).getMenuItemTexts();
```

labkey-danield replied:
Removed redundant code.

labkey-tchad reviewed (commented):
I'd consider finding breadcrumb nodes by index instead of by text (or just alternative methods to do so). This would get real wonky if there are any duplicate values in the path.

labkey-danield committed:
Merge branch 'develop' into fb_fmBoxDisplayAutomation

labkey-danield committed:
Merge branch 'develop' into fb_fmBoxDisplayAutomation

labkey-danield committed:
Integrate Breadcrumb component into FM pages.
Move components up to the StorageView page (remove confusing and unnecessary object).
Remove unused components.
Create the box-display test component.
Use box display in the SMSampleFreezerManagerTest.

labkey-danield commented:
This PR is no longer valid. The code has been moved from TestAutomation to Inventory.
